src/Network/NWEstablishmentReport.cs

```diff
@@ -0,0 +1,135 @@
+
+namespace Network {
+
+	[TV (13,0), Mac (10,15), iOS (13,0), Watch (6,0)]
+	public enum NWReportResolutionSource {
+		Query = 1,
+		Cache = 2,
+		ExpiredCache = 3,
+	}
+
+	[TV (13,0), Mac (10,15), iOS (13,0), Watch (6,0)]
+	public class NWEstablishmentReport : NativeObject {
+
+		public NWEstablishmentReport (IntPtr handle, bool owns) : base (handle, owns) {}
+
+		[DllImport (Constants.NetworkLibrary)]
+		static extern bool nw_establishment_report_get_used_proxy (OS_nw_establishment_report report);
+
+		public bool UsedProxy => nw_establishment_report_get_used_proxy (GetCheckedHandle ());
+
+		[DllImport (Constants.NetworkLibrary)]
+		static extern bool nw_establishment_report_get_proxy_configured (OS_nw_establishment_report report);
+
+		public bool ProxyConfigured => nw_establishment_report_get_proxy_configured (GetCheckedHandle ());
+
+		[DllImport (Constants.NetworkLibrary)]
+		static extern uint nw_establishment_report_get_previous_attempt_count (OS_nw_establishment_report report);
+
+		public uint PreviousAttemptCount => nw_establishment_report_get_previous_attempt_count (GetCheckedHandle ());
+
+		[DllImport (Constants.NetworkLibrary)]
+		static extern ulong nw_establishment_report_get_duration_milliseconds (OS_nw_establishment_report report);
+
+		public ulong Duration => nw_establishment_report_get_duration_milliseconds (GetCheckedHandle ());
+
+		[DllImport (Constants.NetworkLibrary)]
+		static extern ulong nw_establishment_report_get_attempt_started_after_milliseconds (OS_nw_establishment_report report);
+
+		public ulong ConnectionSetupTime => nw_establishment_report_get_attempt_started_after_milliseconds (GetCheckedHandle ());
+
+		[DllImport (Constants.NetworkLibrary)]
+		unsafe static extern void nw_establishment_report_enumerate_resolutions (OS_nw_establishment_report report, void *enumerate_block);
+
+		delegate void nw_report_resolution_enumerator_t (IntPtr block, NWReportResolutionSource source, nuint milliseconds, int endpoint_count, nw_endpoint_t successful_endpoint, nw_endpoint_t preferred_endpoint);
+		static nw_report_resolution_enumerator_t static_ResolutionEnumeratorHandler = TrampolineResolutionEnumeratorHandler;
+
+		[MonoPInvokeCallback (typeof (nw_report_resolution_enumerator_t))]
+		static void TrampolineResolutionEnumeratorHandler (IntPtr block, NWReportResolutionSource source, nuint milliseconds, int endpoint_count, nw_endpoint_t successful_endpoint, nw_endpoint_t preferred_endpoint)
+		{
+			var del = BlockLiteral.GetTarget<Action<NWReportResolutionSource, nuint, int, NWEndpoint, NWEndpoint>> (block);
+			if (del != null) {
+				var nwSuccesfulEndpoint = new NWEndpoint (successful_endpoint, owns: false);
+				var nwPreferredEndpoint = new NWEndpoint (preferred_endpoint, owns: false);
+				del (source, milliseconds, endpoint_count, nwSuccesfulEndpoint, nwPreferredEndpoint);
+			}
+		}
+
+		public void ResolutionEnumeration (Action<NWReportResolutionSource, nuint, int, NWEndpoint, NWEndpoint> handler)
```

Comment:
Better to be named `EnumerateResolutions` I would also be happier if this was an `async` or if there were an `async` version.